rfc/011-registry-authentication.md

```diff
@@ -0,0 +1,281 @@
+
+## Design
+
+The overall idea of this design is to supply an image puller configuration to the VM that's _not measured_.
+Reason for this is that the image pull configuration is a workload owner secret that shouldn't be visible to (unauthenticated) verifiers - see [Alternatives considered](#alternatives-considered).
+
+### Changes to the `imagepuller`
+
+We add support for an optional configuration file to the `imagepuller`, with the following structure:
+
+```go
+import "github.com/google/go-containerregistry/pkg/authn"
+
+type InsecureConfig struct {
+    Registries map[string]Registry // mapping of domain name patterns to registry configuration
+    ExtraEnv map[string]string
+}
+
+type Registry struct {
+    authn.AuthConfig
+    CACerts string // concatenated, PEM-encoded CA certificates (system certs if nil)
+    InsecureSkipVerify bool
+}
+```
+
+On startup, the `imagepuller` reads the config file from `/run/insecure-cfg/imagepuller.TBD`.
+Each key-value pair in `ExtraEnv` is added to the `imagepuller`s own environment.
+The `Registries` are stored in a field.
+
+When a `PullImage` request comes in, the `imagepuller` extracts the registry part from the image reference.
+It then matches the incoming registry to the keys of `Registries`, using the [algorithm in the appendix](#registry-name-matching).
+
+If the `Registry` contains an [`authn.AuthConfig`], it's wrapped with [`authn.FromConfig`] and passed as an option [`remote.WithAuth`].
+The imagepuller constructs a new [`tls.Config`].
+If `CACerts` is not empty, it replaces the pool with a new pool using those certificates.
+If `InsecureSkipVerify` is set, the same is set on the [`tls.Config`].
+The [`tls.Config`] and [`http.ProxyFromEnvironment`] are used to construct an [`http.Transport`] that's passed as an option [`remote.WithTransport`].
+
+[`tls.Config`]: https://pkg.go.dev/crypto/tls#Config
+[`http.ProxyFromEnvironment`]: https://pkg.go.dev/net/http#ProxyFromEnvironment
+[`http.Transport`]: https://pkg.go.dev/net/http#Transport
+[`authn.AuthConfig`]: https://pkg.go.dev/github.com/google/go-containerregistry/pkg/authn#AuthConfig
+[`authn.FromConfig`]: https://pkg.go.dev/github.com/google/go-containerregistry/pkg/authn#FromConfig
+[`remote.WithAuth`]: https://pkg.go.dev/github.com/google/go-containerregistry/pkg/v1/remote#WithAuth
+[`remote.WithTransport`]: https://pkg.go.dev/github.com/google/go-containerregistry/pkg/v1/remote#WithTransport
+
+#### Example 1
+
+In this scenario, there's an internal registry at `registry.corp` for workload images, but `ghcr.io` is used for Contrast images.
+The internal registry is served from within the corporate firewall, while access to external registries is mediated by a proxy.
+Other registries are allowed, as long as they don't require authentication.
+
+```toml
+[extra-env]
+HTTP_PROXY = "https://proxy.corp"
+HTTPS_PROXY = "https://proxy.corp"
+NO_PROXY = ".corp"
+
+[registries."registry.corp."]
```

Comment:
What is the purpose of the leading `.` in `[registries."registry.corp."]`? If this is meant as a "glob everything", I'd rather we use `[registries."registry.corp.*"]` (as well as `[registries."*"]`, and so on, as there's quite some confusion with the toml key notation to be had.

Reply:
The trailing `.` makes this a _fully-qualified domain name_. As mentioned in the appendix, it does not have to be in here, but should be supported for completeness.

The leading `.` indicates "only subdomains should match". We can replace the leading empty label (left of the leading `.`, if you will) with a single asterisk, if that helps readability. Semantics should be the same. What I like about the notation without asterisk is that it does the right thing for the fully-qualified root domain, i.e. `.`.

Reply:
> The trailing `.` makes this a _fully-qualified domain name_. As mentioned in the appendix, it does not have to be in here, but should be supported for completeness.

I'm sorry, I had not seen the new appendix. That would indeed have clarified this.

With this explanation, the `.` makes sense, no need to switch to `*`.

To summarize: sort registries by `key` length in descending order, use first registry for which `imageRefRegistry` ends in`key[0:-1]` if ending in `.`, `key` otherwise. Seems like a nice and simple solution!